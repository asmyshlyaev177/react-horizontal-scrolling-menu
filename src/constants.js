export const rootClassName = 'react-horizontal-scroll-menu'
export const separatorClassName = 'separator'
export const itemClassName = 'item'
export const menuItemsClassName = 'menu-items'
