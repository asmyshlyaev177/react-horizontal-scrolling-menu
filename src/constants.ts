export const rootClassName = 'react-horizontal-scrolling-menu';

export const separatorString = '-separator';

export const separatorClassName = `${rootClassName}-${separatorString}`;
export const itemClassName = `${rootClassName}--item`;
export const scrollContainerClassName = `${rootClassName}--scroll-container`;
export const wrapperClassName = `${rootClassName}--wrapper`;

export const id = 'itemId';

export const dataKeyAttribute = 'data-key';
export const dataIndexAttribute = 'data-index';
