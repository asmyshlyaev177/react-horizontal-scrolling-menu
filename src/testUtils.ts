type IntersectionObserverCB = (arg1: IntersectionObserverEntry[]) => void;

export type { IntersectionObserverCB };

export class MockedObserver {
  cb: IntersectionObserverCB;
  options: IntersectionObserverInit;
  elements: HTMLElement[];

  constructor(cb: IntersectionObserverCB, options: IntersectionObserverInit) {
    this.cb = cb;
    this.options = options;
    this.elements = [];
  }

  unobserve(elem: HTMLElement): void {
    this.elements = this.elements.filter((en) => en !== elem);
  }

  observe(elem: HTMLElement): void {
    this.elements = [...new Set(this.elements.concat(elem))];
  }

  disconnect(): void {
    this.elements = [];
  }

  fire(arr: IntersectionObserverEntry[]): void {
    this.cb(arr);
  }
}

export function traceMethodCalls(obj: object | Function, calls: any = {}) {
  const handler: ProxyHandler<object | Function> = {
    get(target, propKey, receiver) {
      const targetValue = Reflect.get(target, propKey, receiver);
      if (typeof targetValue === 'function') {
        return function (...args: any[]) {
          calls[propKey] = (calls[propKey] || []).concat(args);
          // eslint-disable-next-line @typescript-eslint/ban-ts-comment
          // @ts-ignore
          return targetValue.apply(this, args);
        };
      } else {
        return targetValue;
      }
    },
  };
  return new Proxy(obj, handler);
}
