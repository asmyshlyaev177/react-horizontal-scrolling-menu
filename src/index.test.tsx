import React from 'react';
import { render } from '@testing-library/react';
import { ScrollMenu } from '.';
import { MockedObserver, traceMethodCalls } from './testUtils';
import type { IntersectionObserverCB } from './testUtils';
import * as constants from './constants';

const items = ['test1', 'test2'];
const children = items.map((item) => {
  const itemId = { itemId: item };
  return (
    <div data-testid={item} key={item} {...itemId}>
      {item}
    </div>
  );
});

const setup = ({ ...props } = {}) => {
  return render(<ScrollMenu {...props}>{children}</ScrollMenu>);
};

describe('ScrollMenu', () => {
  let observer: any;
  let mockedObserverCalls: { [k: string]: any } = {};
  beforeEach(() => {
    Object.defineProperty(window, 'IntersectionObserver', {
      writable: true,
      value: jest
        .fn()
        .mockImplementation(function TrackMock(
          cb: IntersectionObserverCB,
          options: IntersectionObserverInit
        ) {
          observer = traceMethodCalls(
            new MockedObserver(cb, options),
            mockedObserverCalls
          );

          return observer;
        }),
    });
  });
  afterEach(() => {
    observer = null;
    mockedObserverCalls = {};
  });

  test('should render without props', () => {
    const { container } = setup();

    expect(container.firstChild).toBeTruthy();
  });

  test('should pass classNames', () => {
    const itemClassName = 'item-class';
    const separatorClassName = 'sep-class';
    const scrollContainerClassName = 'scroll-class';
    const wrapperClassName = 'wrapper-class';

    const { container, debug } = setup({
      itemClassName,
      separatorClassName,
      scrollContainerClassName,
      wrapperClassName,
    });

    debug(container);

    const wrapper = container.firstChild as HTMLElement;
    expect(wrapper.getAttribute('class')).toEqual(
      `${constants.wrapperClassName} ${wrapperClassName}`
    );

    const scrollContainer = wrapper.firstChild as HTMLElement;
    expect(scrollContainer.getAttribute('class')).toEqual(
      `${constants.scrollContainerClassName} ${scrollContainerClassName}`
    );

    const item = scrollContainer.firstChild as HTMLElement;
    expect(item.getAttribute('class')).toEqual(
      `${constants.itemClassName} ${itemClassName}`
    );

    const separator = scrollContainer.childNodes[1] as HTMLElement;
    expect(separator.getAttribute('class')).toEqual(
      `${constants.separatorClassName} ${separatorClassName}`
    );
  });
});
