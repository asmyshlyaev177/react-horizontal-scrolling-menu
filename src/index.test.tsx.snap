// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScrollMenu Children, arrows, header and footer Header and footer 1`] = `
<div>
  <div
    class="react-horizontal-scrolling-menu--wrapper "
  >
    <div
      class="react-horizontal-scrolling-menu--header"
    >
      <div
        data-testid="header"
      >
        Header
      </div>
    </div>
    <div
      class="react-horizontal-scrolling-menu--inner-wrapper"
    >
      <div
        class="react-horizontal-scrolling-menu--arrow-left"
      />
      <div
        class="react-horizontal-scrolling-menu--scroll-container scroll-class"
      >
        <div
          class="react-horizontal-scrolling-menu--item "
          data-index="0"
          data-key="test1"
        >
          <div
            data-testid="test1"
          >
            {"itemId":"test1","isFirstItemVisible":false,"isLastItemVisible":false,"visibleElements":["test1","test2"],"visibleElementsWithSeparators":["test1","item1-separator","test2"],"visibleItems":["test1","item1-separator","test2"],"visibleItemsWithoutSeparators":["test1","test2"],"initComplete":true}
          </div>
        </div>
        <div
          class="react-horizontal-scrolling-menu--separator "
          data-index="0.1"
          data-key="test1-separator"
        />
        <div
          class="react-horizontal-scrolling-menu--item "
          data-index="1"
          data-key="test2"
        >
          <div
            data-testid="test2"
          >
            {"itemId":"test2","isFirstItemVisible":false,"isLastItemVisible":false,"visibleElements":["test1","test2"],"visibleElementsWithSeparators":["test1","item1-separator","test2"],"visibleItems":["test1","item1-separator","test2"],"visibleItemsWithoutSeparators":["test1","test2"],"initComplete":true}
          </div>
        </div>
      </div>
      <div
        class="react-horizontal-scrolling-menu--arrow-right"
      />
    </div>
    <div
      class="react-horizontal-scrolling-menu--footer"
    >
      <div
        data-testid="footer"
      >
        Footer
      </div>
    </div>
  </div>
</div>
`;

exports[`ScrollMenu Children, arrows, header and footer LeftArrow, ScrollContainer, MenuItems, RightArrow 1`] = `
<div>
  <div
    class="react-horizontal-scrolling-menu--wrapper "
  >
    <div
      class="react-horizontal-scrolling-menu--header"
    />
    <div
      class="react-horizontal-scrolling-menu--inner-wrapper"
    >
      <div
        class="react-horizontal-scrolling-menu--arrow-left"
      >
        <div
          class="left-arrow"
          data-testid="left-arrow"
        >
          {"isFirstItemVisible":false,"isLastItemVisible":false,"visibleElements":["test1","test2"],"visibleElementsWithSeparators":["test1","item1-separator","test2"],"visibleItems":["test1","item1-separator","test2"],"visibleItemsWithoutSeparators":["test1","test2"],"initComplete":true}
        </div>
      </div>
      <div
        class="react-horizontal-scrolling-menu--scroll-container scroll-class"
      >
        <div
          class="react-horizontal-scrolling-menu--item "
          data-index="0"
          data-key="test1"
        >
          <div
            data-testid="test1"
          >
            {"itemId":"test1","isFirstItemVisible":false,"isLastItemVisible":false,"visibleElements":["test1","test2"],"visibleElementsWithSeparators":["test1","item1-separator","test2"],"visibleItems":["test1","item1-separator","test2"],"visibleItemsWithoutSeparators":["test1","test2"],"initComplete":true}
          </div>
        </div>
        <div
          class="react-horizontal-scrolling-menu--separator "
          data-index="0.1"
          data-key="test1-separator"
        />
        <div
          class="react-horizontal-scrolling-menu--item "
          data-index="1"
          data-key="test2"
        >
          <div
            data-testid="test2"
          >
            {"itemId":"test2","isFirstItemVisible":false,"isLastItemVisible":false,"visibleElements":["test1","test2"],"visibleElementsWithSeparators":["test1","item1-separator","test2"],"visibleItems":["test1","item1-separator","test2"],"visibleItemsWithoutSeparators":["test1","test2"],"initComplete":true}
          </div>
        </div>
      </div>
      <div
        class="react-horizontal-scrolling-menu--arrow-right"
      >
        <div
          class="right-arrow"
          data-testid="right-arrow"
        >
          {"isFirstItemVisible":false,"isLastItemVisible":false,"visibleElements":["test1","test2"],"visibleElementsWithSeparators":["test1","item1-separator","test2"],"visibleItems":["test1","item1-separator","test2"],"visibleItemsWithoutSeparators":["test1","test2"],"initComplete":true}
        </div>
      </div>
    </div>
    <div
      class="react-horizontal-scrolling-menu--footer"
    />
  </div>
</div>
`;

exports[`ScrollMenu should render without props 1`] = `
<div>
  <div
    class="react-horizontal-scrolling-menu--wrapper "
  >
    <div
      class="react-horizontal-scrolling-menu--header"
    />
    <div
      class="react-horizontal-scrolling-menu--inner-wrapper"
    >
      <div
        class="react-horizontal-scrolling-menu--arrow-left"
      />
      <div
        class="react-horizontal-scrolling-menu--scroll-container "
      >
        <div
          class="react-horizontal-scrolling-menu--item "
          data-index="0"
          data-key="test1"
        >
          <div
            data-testid="test1"
          >
            {"itemId":"test1","isFirstItemVisible":false,"isLastItemVisible":false,"visibleElements":["test1","test2"],"visibleElementsWithSeparators":["test1","item1-separator","test2"],"visibleItems":["test1","item1-separator","test2"],"visibleItemsWithoutSeparators":["test1","test2"],"initComplete":true}
          </div>
        </div>
        <div
          class="react-horizontal-scrolling-menu--separator "
          data-index="0.1"
          data-key="test1-separator"
        />
        <div
          class="react-horizontal-scrolling-menu--item "
          data-index="1"
          data-key="test2"
        >
          <div
            data-testid="test2"
          >
            {"itemId":"test2","isFirstItemVisible":false,"isLastItemVisible":false,"visibleElements":["test1","test2"],"visibleElementsWithSeparators":["test1","item1-separator","test2"],"visibleItems":["test1","item1-separator","test2"],"visibleItemsWithoutSeparators":["test1","test2"],"initComplete":true}
          </div>
        </div>
      </div>
      <div
        class="react-horizontal-scrolling-menu--arrow-right"
      />
    </div>
    <div
      class="react-horizontal-scrolling-menu--footer"
    />
  </div>
</div>
`;
